votingpool/input_selection.go

```diff
@@ -0,0 +1,286 @@
+	if c[i].Address().seriesID < c[j].Address().seriesID {
+		return true
+	}
+
+	if c[i].Address().seriesID == c[j].Address().seriesID &&
+		c[i].Address().index < c[j].Address().index {
+		return true
+	}
+
+	if c[i].Address().seriesID == c[j].Address().seriesID &&
+		c[i].Address().index == c[j].Address().index &&
+		c[i].Address().branch < c[j].Address().branch {
+		return true
+	}
+
+	txidComparison := bytes.Compare(c[i].TxSha().Bytes(), c[j].TxSha().Bytes())
+
+	if c[i].Address().seriesID == c[j].Address().seriesID &&
+		c[i].Address().index == c[j].Address().index &&
+		c[i].Address().branch == c[j].Address().branch &&
+		txidComparison < 0 {
+		return true
+	}
+
+	if c[i].Address().seriesID == c[j].Address().seriesID &&
+		c[i].Address().index == c[j].Address().index &&
+		c[i].Address().branch == c[j].Address().branch &&
+		txidComparison == 0 &&
+		c[i].OutputIndex() < c[j].OutputIndex() {
+		return true
+	}
+	return false
+}
+
+// getEligibleInputs returns eligible inputs with addresses between startAddress
+// and the last used address of lastSeriesID.
+func (p *Pool) getEligibleInputs(
+	store *txstore.Store,
+	startAddress WithdrawalAddress,
+	lastSeriesID uint32,
+	dustThreshold btcutil.Amount,
+	chainHeight int32,
+	minConf int) ([]Credit, error) {
+
+	if p.GetSeries(lastSeriesID) == nil {
+		str := fmt.Sprintf("lastSeriesID (%d) does not exist", lastSeriesID)
+		return nil, newError(ErrSeriesNotExists, str, nil)
+	}
+	unspents, err := store.UnspentOutputs()
+	if err != nil {
+		return nil, newError(ErrInputSelection, "failed to get unspent outputs", err)
+	}
+	addrMap, err := groupCreditsByAddr(unspents, p.manager.ChainParams())
+	if err != nil {
+		return nil, err
+	}
+	var inputs []Credit
+	address := startAddress
+	for {
+		log.Debugf("Looking for eligible inputs at address %v", address.addrIdentifier())
+		if candidates, ok := addrMap[address.addr.EncodeAddress()]; ok {
+			var eligibles []Credit
+			for _, c := range candidates {
+				if p.isCreditEligible(c, minConf, chainHeight, dustThreshold) {
+					eligibles = append(eligibles, newCredit(c, address))
+				}
+			}
+			// Make sure the eligibles are correctly sorted.
+			sort.Sort(byAddress(eligibles))
+			inputs = append(inputs, eligibles...)
+		}
+		nAddr, err := nextAddr(p, address.seriesID, address.branch, address.index, lastSeriesID+1)
+		if err != nil {
+			return nil, newError(ErrInputSelection, "failed to get next withdrawal address", err)
+		} else if nAddr == nil {
+			log.Debugf("getEligibleInputs: reached last addr, stopping")
+			break
+		}
+		address = *nAddr
+	}
+	return inputs, nil
+}
+
+// nextAddr returns the next WithdrawalAddress according to the input selection
+// rules: http://opentransactions.org/wiki/index.php/Input_Selection_Algorithm_(voting_pools)
+// It returns nil if the new address' seriesID is >= stopSeriesID.
+func nextAddr(p *Pool, seriesID uint32, branch Branch, index Index, stopSeriesID uint32) (
+	*WithdrawalAddress, error) {
+	branch++
```

Comment:
Probably doesn't matter much sense it's a local, but shouldn't this mutation happen after the attempt to get the series?  That is to say, if the series doesn't exist, there is no reason to increment the branch.